tests/contrib/test_gce.py

```diff
@@ -14,102 +14,195 @@
+        except ValueError as e:
+            error = e
+        self.assertIsNotNone(error)
 
+    @mock.patch(METADATA_SERVER, return_value=A_SERVICE_ACCOUNT)
+    def test_default_service_account_info(self, get_metadata):
+        credentials = AppAssertionCredentials()
+        self.assertEquals(
+            credentials._service_account_info,
+            DEFAULT_SERVICE_ACCOUNT
+        )
+        self.assertEquals(credentials.scopes, A_SERVICE_ACCOUNT['scopes'])
+        self.assertEquals(
+            credentials.service_account_email,
+            A_SERVICE_ACCOUNT['email']
+        )
+        self.assertEquals(credentials.service_account_info, A_SERVICE_ACCOUNT)
+        get_metadata.assert_called_once_with(
+            path=[
+                'instance',
+                'service-accounts',
+                'default'
+            ],
+            http_request=None
+        )
+
+    @mock.patch(METADATA_SERVER, return_value=A_SERVICE_ACCOUNT)
+    def test_custom_service_account_info(self, get_metadata):
+        credentials = AppAssertionCredentials(
+            service_account_email=A_SERVICE_ACCOUNT['email']
+        )
+        self.assertEquals(
+            credentials._service_account_info,
+            {'email': A_SERVICE_ACCOUNT['email']}
+        )
+        self.assertEquals(credentials.scopes, A_SERVICE_ACCOUNT['scopes'])
+        self.assertEquals(
+            credentials.service_account_email,
+            A_SERVICE_ACCOUNT['email']
+        )
+        self.assertEquals(credentials.service_account_info, A_SERVICE_ACCOUNT)
+        get_metadata.assert_called_once_with(
+            path=[
+                'instance',
+                'service-accounts',
+                A_SERVICE_ACCOUNT['email']
+            ],
+            http_request=None
+        )
+
+    @mock.patch(METADATA_SERVER, return_value='a-project-id')
+    def test_project_id(self, get_metadata):
+        credentials = AppAssertionCredentials()
+        self.assertIsNone(credentials._project_id)
+        self.assertEquals(credentials.project_id, 'a-project-id')
+        self.assertEquals(credentials.project_id, 'a-project-id')
+        get_metadata.assert_called_once_with(
+            path=[
+                'project',
+                'project-id'
+            ],
+            recursive=False
+        )
+
+    def test_refresh_token(self):
         credentials = AppAssertionCredentials()
         self.assertEquals(None, credentials.access_token)
-        credentials.refresh(http)
-        self.assertEquals(access_token, credentials.access_token)
 
-        base_metadata_uri = (
-            'http://metadata.google.internal/computeMetadata/v1/instance/'
-            'service-accounts/default/token')
-        http.request.assert_called_once_with(
-            base_metadata_uri, headers={'Metadata-Flavor': 'Google'})
+        with mock.patch(METADATA_SERVER, return_value=ONE_SECOND_TOKEN):
+            credentials.get_access_token()
+            time.sleep(2)
```

Comment:
That doesn't work, because `datetime` is immutable. Mocking `datetime.datetime.now()` appears to be non-trivial, there are some libraries that do it well. Thoughts on making one of them a test dependency?